### PR TITLE
Redux setup

### PR DESCRIPTION
**Pull request by hernan-clich**
## Summary

Added Redux to the project so the book list is now obtained using the `useSelector` hook from `react-redux` instead of using it directly from the mock file.
Added the last remaining piece of config to link Redux with Reactotron.

### Failure case

redux_failure

### Success case

redux_success

## Trello Card

https://trello.com/c/g9xojyC7/25-n-91-redux-i-actions-reducers-store

### CODE REVIEW

File: wbooks/App.test.tsx [hunk 1]
```diff
@@ -5,11 +5,22 @@
 import 'react-native';
 import React from 'react';
 import renderer from 'react-test-renderer';
+import { Provider } from 'react-redux';
+import configureMockStore from 'redux-mock-store';
 
 import App from './src/app';
 
+const mockStore = configureMockStore([]);
```
Comment (sfernandez11) [inline wbooks/App.test.tsx]: `const mockStore = configureMockStore([thunk]);`

File: wbooks/App.test.tsx [hunk 2]
```diff
@@ -5,11 +5,22 @@
...
+const store = mockStore({
+  bookReducer: []
```
Comment (sfernandez11) [inline wbooks/App.test.tsx]: `books: { books: [] }`

File: wbooks/src/app/screens/Library/index.tsx [hunk 3]
```diff
@@ -15,6 +16,8 @@ interface Props {
 }
 
 function Library({ navigation }: Props) {
+  const books: Book[] = useSelector((state: RootState) => state.books);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: `const books = useSelector<RootState, Book[]>(state => state.books.books);`

File: wbooks/src/app/screens/Library/index.tsx [hunk 4]
```diff
@@ -15,6 +16,8 @@ interface Props {
...
+
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx:28]: You must use the `useEffect` + `dispatch` to get and set the books in the reducer

File: wbooks/src/app/screens/Library/index.tsx [hunk 5]
```diff
@@ -28,12 +31,7 @@ function Library({ navigation }: Props) {
 
   return (
     <SafeAreaView style={styles.container}>
-      <FlatList
-        style={styles.container}
-        data={BOOKS_MOCK}
-        renderItem={renderBooks}
-        keyExtractor={keyExtractor}
-      />
+      <FlatList style={styles.container} data={books} renderItem={renderBooks} keyExtractor={keyExtractor} />
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Is it necessary to have the same style in the `SafeAreaView` and `FlatList`? I'm think not

File: wbooks/src/redux/book/actions.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Book } from '@interfaces/books';
+
+interface ActionTypes {
+  GET_BOOKS: '@@BOOKS/GET_BOOKS';
+}
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Not necessary

File: wbooks/src/redux/book/actions.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+export const actions: ActionTypes = {
+  GET_BOOKS: '@@BOOKS/GET_BOOKS'
+};
+
+function getBooks(books: Book[]) {
+  return {
+    type: actions.GET_BOOKS,
+    books
+  };
+}
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Wrong implementation

File: wbooks/src/redux/book/reducer.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { AnyAction } from 'redux';
+import { BOOKS_MOCK } from '@constants/mockBooks';
+import { Book } from '@interfaces/books';
+
+import { actions } from './actions';
+
+const initialState: Book[] = BOOKS_MOCK;
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: Check this `initialState`, each reducer must be an object
```
const initialState: BookState = {
 books: []
};
```

Create `BookState` interface. Also, add in this PR the `SUCCESS`  and `FAILURE` cases too

File: wbooks/src/redux/book/actions.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+export const actions: ActionTypes = {
+  GET_BOOKS: '@@BOOKS/GET_BOOKS'
+};
+
+function getBooks(books: Book[]) {
+  return {
+    type: actions.GET_BOOKS,
+    books
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Wrong Action name, you have Action type and Action payload

File: wbooks/src/redux/book/reducer.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+function booksReducer(state = initialState, action: AnyAction) {
+  switch (action.type) {
+    case actions.GET_BOOKS:
+      return action.books;
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: Update this too

File: wbooks/src/redux/store.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+import thunk from 'redux-thunk';
+import Reactotron from '@config/reactotronConfig';
+
+import booksReducer from './book/reducer';
+
+const rootReducer = combineReducers({
+  books: booksReducer
+});
+
+export type RootState = ReturnType<typeof rootReducer>;
```
Comment (sfernandez11) [inline wbooks/src/redux/store.ts]: Move this to an `interface` file, remember the folders structure

File: wbooks/src/redux/book/actions.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+interface ActionTypes {
+  GET_BOOKS: '@@BOOKS/GET_BOOKS';
+}
+
+export const actions: ActionTypes = {
+  GET_BOOKS: '@@BOOKS/GET_BOOKS'
+};
+
+function getBooks(books: Book[]) {
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts]: Add in this PR the `SUCCESS`  and `FAILURE` cases too

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Great, some comments...

File: wbooks/src/app/screens/Library/index.tsx (hunk 5 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx]: You were right, I removed it from the `FlatList` and there's no difference.

File: wbooks/src/app/screens/Library/index.tsx [hunk 13]
```diff
@@ -16,6 +19,13 @@ interface Props {
 }
 
 function Library({ navigation }: Props) {
+  const dispatch = useDispatch();
+  const { books, error } = useSelector<RootState, BookState>(state => state.books);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Why the `error`?

File: wbooks/src/app/screens/Library/index.tsx [hunk 14]
```diff
@@ -16,6 +19,13 @@ interface Props {
...
+
+  React.useEffect(() => {
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Extract `useEffect` from React

File: wbooks/src/app/screens/Library/index.tsx [hunk 15]
```diff
@@ -16,6 +19,13 @@ interface Props {
...
+    dispatch(actionCreators.getBooks());
+  }, [dispatch, books, error]);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Why `books` and `error` are dependencies?

File: wbooks/src/redux/book/reducer.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+};
+
+function booksReducer(state = initialState, action: AnyAction) {
+  switch (action.type) {
+    case actions.GET_BOOKS:
+      return state;
+    case actions.GET_BOOKS_SUCCESS:
+      return { ...state, books: action.payload };
+    case actions.GET_BOOKS_FAILURE:
+      return { ...state, error: action.payload };
```
Comment (sfernandez11) [inline wbooks/src/redux/book/reducer.ts]: You forgot the `loading`. While you are fetching some data you will have some delay time, use a `loading` value for this

**Review by sfernandez11 — COMMENTED** [review]
Last comments...

File: wbooks/src/app/screens/Library/index.tsx (hunk 14 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx]: Ok

File: wbooks/src/app/screens/Library/index.tsx (hunk 13 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx]: Removed

File: wbooks/src/redux/book/reducer.ts (hunk 16 above)
Comment (hernan-clich) [reply wbooks/src/redux/book/reducer.ts]: Added

File: wbooks/src/app/screens/Library/index.tsx (hunk 15 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx]: Both removed

**Comment by hernan-clich** [comment]
Refactored according to your suggestions and added screenshots for both success and failure cases.

**Review by sfernandez11 — DISMISSED** [review]
🥇

**Review by sfernandez11 — APPROVED** [review]

**Pull request merged**